{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fa1bbf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql as sql\n",
    "\n",
    "f = open('gene.txt',\"r\")\n",
    "content = f.readlines()\n",
    "f.close()\n",
    "\n",
    "gene_symbol = content[0].split(\" \")[0]\n",
    "gene_id = content[1].split(\",\")[0].split(\":\")[1].strip()\n",
    "print (gene_symbol)\n",
    "print (gene_id)\n",
    "\n",
    "indx_lst = [i for i in range(len(content)) if \"mRNA transcript\" in content[i]]\n",
    "\n",
    "connection = sql.connect(user='root',password=\"987654321\",host='localhost',port=3306)\n",
    "db = connection.cursor()\n",
    "def get_ex_dt(data_content):\n",
    "    rt = {}\n",
    "    ex_no = 1\n",
    "    for dt in data_content:\n",
    "        ex_dt = dt.split(\"\\t\")[0]\n",
    "        ex_zt = int(ex_dt.split(\"-\")[0])\n",
    "        ex_ed = int(ex_dt.split(\"-\")[1])\n",
    "        ex_sz = ex_ed - ex_zt + 1 if ex_ed > ex_zt else ex_zt - ex_ed + 1\n",
    "        rt[ex_no] = {\"st\":ex_zt, \"ed\":ex_ed, \"sz\":ex_sz}\n",
    "        ex_no += 1\n",
    "    return rt\n",
    "\n",
    "def create_db(db_name):\n",
    "    db.execute(\"create database if not exists %s\"%(db_name))\n",
    "    connection.commit()\n",
    "    \n",
    "def create_table(db_name,table_name):\n",
    "    db.execute(\"use %s\"%(db_name))\n",
    "    db.execute(\"create table if not exists %s(gene_symbol varchar(100), gene_id varchar(100), trans_id varchar(100), exon_no varchar(100), exon_st varchar(100), exon_ed varchar(100), exon_sz varchar(100))\"%(table_name))\n",
    "\n",
    "def upload_data_to_table(gene_symbol,gene_id,tran_id,exon_no,exon_st,exon_ed,exon_sz,db_name,table_name):\n",
    "    db.execute(\"use %s\"%(db_name))\n",
    "    db.execute(f\"insert into {table_name} (gene_symbol,gene_id,trans_id,exon_no,exon_st,exon_ed,exon_sz) values ('{gene_symbol}','{gene_id}', '{tran_id}', '{exon_no}', '{exon_st}', '{exon_ed}', '{exon_sz}')\")\n",
    "    connection.commit()\n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "    create_db(\"PYDT_VLCY_PRJ\")\n",
    "    create_table(\"PYDT_VLCY_PRJ\", \"GENE_TABLE\")\n",
    "    db_name = \"PYDT_VLCY_PRJ\"\n",
    "    table_name = \"GENE_TABLE\"\n",
    "    with open(\"gene_table.csv\", \"w\") as fl:\n",
    "        fl.write(\"Gene symbol,Gene ID,Trans ID,Exon No,Exon st,Exon ed,Exon sz\")\n",
    "        for i in range(len(indx_lst)):\n",
    "            if i != len(indx_lst) - 1:\n",
    "                st,ed = indx_lst[i],indx_lst[i+1]\n",
    "                data_content = content[st:ed]\n",
    "                tr_id = [x for x in data_content[0].split(\" \") if \"NM_\" in x][0]\n",
    "                tr_id = tr_id.replace(\",\",\"\")\n",
    "                \n",
    "                idx = [i for i in range(len(data_content)) if \"--------------\" in data_content[i]][0]\n",
    "                data_content = data_content[idx+1:]\n",
    "                data_content = [x.strip() for x in data_content if x.strip() != '']\n",
    "                \n",
    "                f_d = get_ex_dt(data_content)\n",
    "                for ex in f_d:\n",
    "                    fl.write(\"\\n%s,%s,%s,%s,%s,%s,%s\"%(gene_symbol, gene_id, tr_id, ex, f_d[ex][\"st\"],f_d[ex][\"ed\"],f_d[ex][\"sz\"]))\n",
    "                    upload_data_to_table(gene_symbol,gene_id,tr_id, ex, f_d[ex][\"st\"],f_d[ex][\"ed\"],f_d[ex][\"sz\"],db_name,table_name)\n",
    "                print (tr_id)\n",
    "                \n",
    "            else:\n",
    "                st = indx_lst[i]\n",
    "                data_content = content[st:]\n",
    "                tr_id = [x for x in data_content[0].split(\" \") if \"NM_\" in x][0]\n",
    "                tr_id = tr_id.replace(\",\",\"\")\n",
    "                \n",
    "                idx = [i for i in range(len(data_content)) if \"--------------\" in data_content[i]][0]\n",
    "                data_content = [x.strip() for x in data_content if x.strip() != '']\n",
    "                data_content = data_content[idx+1:]\n",
    "                \n",
    "                f_d = get_ex_dt(data_content)\n",
    "                for ex in f_d:\n",
    "                    fl.write(\"\\n%s,%s,%s,%s,%s,%s,%s\"%(gene_symbol, gene_id, tr_id, ex, f_d[ex][\"st\"],f_d[ex][\"ed\"],f_d[ex][\"sz\"]))\n",
    "                    upload_data_to_table(gene_symbol,gene_id,tr_id, ex, f_d[ex][\"st\"],f_d[ex][\"ed\"],f_d[ex][\"sz\"],db_name,table_name)\n",
    "                print (tr_id)\n",
    "                \n",
    "db.close()\n",
    "connection.close()               "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
